Derive remaining managers from ManagerBase

gassmoeller opened the pull request:
A suggestion for how to finish #1775. This is still work in progress and unfortunately not completely backwards compatible, but at least the incompatibility is only limited to two rarely used functions in the code, not the input file. The traction manager (and the velocity manager once I get to it) need additional information compared to `ManagerBase`, namely which boundary each plugin is responsible for, and which components (of the tensor components xyz). The main change is the type of internal storage container, and the access functions `get_active_boundary_traction_conditions` and `get_active_boundary_traction_names`. Opinions? If we we agree this is a good idea I can continue with the velocity manager.

bangerth commented on source/boundary_traction/interface.cc:
```diff
@@ -78,20 +61,28 @@ namespace aspect
+      Tensor<1,dim> traction;
 
-      Assert(boundary_plugins != boundary_traction_objects.end(),
+      [[maybe_unused]] bool found_plugin = false;
```
We're not there yet -- we still allow compiling ASPECT with C++14.

bangerth commented on source/boundary_traction/interface.cc:
```diff
@@ -78,20 +61,28 @@ namespace aspect
 
-      Assert(boundary_plugins != boundary_traction_objects.end(),
+      [[maybe_unused]] bool found_plugin = false;
+      unsigned int i=0;
+      for (const auto &plugin: this->plugin_objects)
+        {
+          if (this->boundary_indicators[i])
+            {
+              const Tensor<1,dim> plugin_traction = plugin->boundary_traction(boundary_indicator,
+                                                                              position,
+                                                                              normal_vector);
+              for (unsigned int d=0; d<dim; ++d)
+                if (this->component_masks[i][d])
+                  traction[d] += plugin_traction[d];
+            }
+        }
```
Something is not right with this code. I don't see you incrementing `i` anywhere. It's also not clear that `this-boundary_indicators[i]` is a `bool`. Are you forgetting to compare it against `boundary_indicator`?

bangerth reviewed (commented):
Yes, nice idea. I supposed you also want to do this for the boundary velocity manager?

gassmoeller commented:
/rebuild

gassmoeller commented:
> Something is not right with this code. I don't see you incrementing i anywhere. It's also not clear that this-boundary_indicators[i] is a bool. Are you forgetting to compare it against boundary_indicator?

Yes, I rushed this PR a bit, I mostly needed feedback if the idea is sound in general. I have cleaned up the PR, addressed your comments and expanded it to the boundary velocity manager. I think this should work now, but I am open to suggestions for changes.

tjhei commented on include/aspect/boundary_traction/interface.h:
```diff
@@ -193,9 +211,37 @@ namespace aspect
          * A list of boundary traction objects that have been requested in the
          * parameter file.
+         *
+         * @deprecated: This variable is no longer used, but needed to issue a proper
```
```suggestion
         * @deprecated This variable is no longer used, but needed to issue a proper
```

tjhei commented on include/aspect/boundary_traction/interface.h:
```diff
@@ -206,6 +252,11 @@ namespace aspect
          * "function"). If the components string is empty, it is assumed the
          * plugins are used for all components.
+         *
+         * @deprecated: Remove this variable when the deprecated functions
```
```suggestion
         * @deprecated Remove this variable when the deprecated functions
```

tjhei reviewed (approved):
minor nitpick for doxygen formatting...

gassmoeller commented:
Thanks for the review. I have addressed all comments, all commits are squashed and I updated the date of the changelog entry. I will leave this open for a day or two and merge if there are no other comments.
